### DISCUSSION TITLE
PEFT and LoRa for MLX?

### DISCUSSION THREAD

**Discussion by pow9791** [ml-explore/mlx#845] [General]
How do I run code for PEFT and LoRa in MLX? I want to fine a mistral model locally.

**Comment by awni** [comment]
Here's some docs on [using LoRA](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/LORA.md) with MLX LM.

**Reply by pow9791** [reply]
It didn't work right, I tried using this: 
```cd mlx-examples/lora
pip install -r requirements.txt 
 python lora.py \
 --train \
 --model mistralai/Mistral-7B-Instruct-v0.2 \
 --data ~/Developer/AI/Models/fine-tuning/data \
 --batch-size 2 \
 --lora-layers 8 \
 --iters 1000 ```, it said like batch size not defined.

**Reply by awni** [reply]
That command doesn't look right to me. Try something like this:

```shell
python -m mlx_lm.lora \
    --model <path_to_model> \
    --adapter-file <path_to_adapters.npz> \
    --data <path_to_data> \
    --batch- size 2
```

**Reply by pow9791** [reply]
Thanks!

**Reply by pow9791** [reply]
Suppose  I can get the fine tuning to work, how do I export a model fine tuned on MLX to run on Pytorch with CUDA ?

**Reply by awni** [reply]
After fine-tuning, [fuse the adaptors into the base model](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/LORA.md#Fuse-and-Upload) which will save it in `safetensors`. From there you should be able to load it and use it with Hugging Face Transformers (which uses PyTorch).

For a more detailed guide check out e.g. https://apeatling.com/articles/part-4-testing-and-interacting-with-your-fine-tuned-llm/

**Reply by pow9791** [reply]
I can't find the adapter.npz under the mlx-examples folder, I used mdfind "adapter.npz" still couldn't find it. Is it the LoRa file? I also I used your code, it gave this error, Error while finding module specification for 'mlx_lm.lora' (ModuleNotFoundError: No module named 'mlx_lm') Can't someone update MPS for pytorch more often?

**Reply by awni** [reply]
>Error while finding module specification for 'mlx_lm.lora' 

python -m pip install -U mlx-lm

>I can't find the adapter.npz under the mlx-examples folder, I used mdfind "adapter.npz" still couldn't find it.

They will be saved wherever you run the command from
